dnssd/src/main/java/com/github/druk/dnssd/DNSSD.java

```diff
@@ -555,30 +571,41 @@ public int reconfirmRecord(int flags, int ifIndex, String fullName, int rrtype,
 
     @Override
     public void onServiceStarting() {
-        if (multicastLock == null) {
-            synchronized (this) { // Double-check lock
-                if (multicastLock == null) {
-                    WifiManager wifi = (WifiManager) context.getApplicationContext()
-                            .getSystemService(Context.WIFI_SERVICE);
-                    if (wifi == null) {
-                        Log.wtf("DNSSD", "Can't get WIFI Service");
-                        return;
+        if(lockEnabled) { // optional enable/disable (default enabled)
+            if (multicastLock == null) {
+                synchronized (this) { // Double-check lock
+                    if (multicastLock == null) {
+                        WifiManager wifi = (WifiManager) context.getApplicationContext()
+                                .getSystemService(Context.WIFI_SERVICE);
+                        if (wifi == null) {
+                            Log.wtf("DNSSD", "Can't get WIFI Service");
+                            return;
+                        }
+                        multicastLock = wifi.createMulticastLock(MULTICAST_LOCK_NAME);
+                        multicastLock.setReferenceCounted(true);
                     }
-                    multicastLock = wifi.createMulticastLock(MULTICAST_LOCK_NAME);
-                    multicastLock.setReferenceCounted(true);
                 }
             }
+            synchronized (this) {
+                // prevent multiple calls to acquire if we have it held
+                // onServiceStarting may be called multiple times
+                // fixes fatal crash on too many multicast locks held
+                if(!multicastLock.isHeld()) multicastLock.acquire();
+            }
         }
-        multicastLock.acquire();
     }
 
     @Override
     public void onServiceStopped() {
-        if (multicastLock == null) {
-            Log.wtf("DNSSD", "Multicast lock doesn't exist");
-            return;
+        if(lockEnabled) {
+            if (multicastLock == null) {
+                Log.wtf("DNSSD", "Multicast lock doesn't exist");
+                return;
+            }
+            synchronized (this) {
+                if (multicastLock.isHeld()) multicastLock.release();
```

Comment:
This will stop all DNS-SD after the first onServiceStopped
Some kind of counter should be here

Reply:
See failed tests